src/SIMD.jl

```diff
@@ -50,6 +50,48 @@ uinttype{T}(::Type{T}) = uinttype(Val{8*sizeof(T)})
 
 =#
 
+# Array types for SIMD
+
+using Base: Slice, ScalarIndex
+
+"""
+    ContiguousSubArray{T,N,P,I,L}
+
+Like `Base.FastContiguousSubArray` but without requirement for linear
+indexing (i.e., type parameter `L` can be `false`).
+
+# Examples
+```
+julia> A = view(ones(5, 5), :, [1,3]);
+
+julia> A isa Base.FastContiguousSubArray
+false
+
+julia> A isa SIMD.ContiguousSubArray
+true
+```
+"""
+ContiguousSubArray{T,N,P,
+                   I<:Union{Tuple{Union{Slice, AbstractUnitRange}, Vararg{Any}},
+                            Tuple{Vararg{ScalarIndex}}},
+                   L} = SubArray{T,N,P,I,L}
+
+"""
+    ContiguousArray{T,N}
+
+Array types with contiguous first dimension.
+"""
+ContiguousArray{T,N} = Union{Array{T,N}, ContiguousSubArray{T,N}}
```

Comment:
```suggestion
ContiguousArray{T,N} = Union{DenseArray{T,N}, ContiguousSubArray{T,N}}
```

This would be just as accurate and allow for a few more subtypes (e.g., `CodeUnits`).